src/lewis/core/control_server.py

```diff
@@ -329,12 +330,15 @@ def process(self):
 
         If the server has not been started yet (via :meth:`start_server`), a RuntimeError
         is raised.
+
+        :param blocking: If True, this function will block until it has received data or a timeout
+                         is triggered. Default is False to preserve behavior of prior versions.
         """
         if self._socket is None:
             raise RuntimeError('The server has not been started yet, use start_server to do so.')
 
         try:
-            request = self._socket.recv_unicode(flags=zmq.NOBLOCK)
+            request = self._socket.recv_unicode(flags=zmq.NOBLOCK if not blocking else 0)
```

Comment:
Using this test script:

```python
from lewis.core.control_client import ControlClient

client = ControlClient(host='127.0.0.1', port='10000')

print 'hello '

for i in range(1000):
    sim = client.get_object('simulation')
    del sim

print 'world'
```

And running it against some arbitrary simulation:

```
$ lewis -r localhost:10000 chopper
$ python test.py
```

Lewis often crashes with this error:
```
2017-09-20 16:29:08,683 INFO lewis.ChopperEpicsInterface.PropertyExposingDriver: Processed PV updates: State=init
2017-09-20 16:29:13,126 WARNING lewis.Simulation: Stopping simulation
2017-09-20 16:29:13,127 INFO lewis.Simulation: Stopped processing control server commands, ending thread.
2017-09-20 16:29:13,128 INFO lewis.AdapterCollection: Disconnecting device interface for protocol 'epics'
Traceback (most recent call last):
  File "/home/mikehart/Dev/lewis/env/bin/lewis", line 11, in <module>
    load_entry_point('lewis', 'console_scripts', 'lewis')()
  File "/home/mikehart/Dev/lewis/src/lewis/scripts/run.py", line 245, in run_simulation
    simulation.start()
  File "/home/mikehart/Dev/lewis/src/lewis/core/simulation.py", line 184, in start
    delta = self._process_cycle(delta)
  File "/home/mikehart/Dev/lewis/src/lewis/core/simulation.py", line 220, in _process_cycle
    self._process_simulation_cycle(delta)
  File "/home/mikehart/Dev/lewis/src/lewis/core/simulation.py", line 247, in _process_simulation_cycle
    self._control_server.process()
  File "/home/mikehart/Dev/lewis/src/lewis/core/control_server.py", line 341, in process
    request = self._socket.recv_unicode(flags=zmq.NOBLOCK if not blocking else 0)
  File "/home/mikehart/Dev/lewis/env/local/lib/python2.7/site-packages/zmq/sugar/socket.py", line 457, in recv_string
    msg = self.recv(flags=flags)
  File "zmq/backend/cython/socket.pyx", line 693, in zmq.backend.cython.socket.Socket.recv (zmq/backend/cython/socket.c:7683)
  File "zmq/backend/cython/socket.pyx", line 727, in zmq.backend.cython.socket.Socket.recv (zmq/backend/cython/socket.c:7460)
  File "zmq/backend/cython/socket.pyx", line 150, in zmq.backend.cython.socket._recv_copy (zmq/backend/cython/socket.c:2437)
  File "zmq/backend/cython/socket.pyx", line 145, in zmq.backend.cython.socket._recv_copy (zmq/backend/cython/socket.c:2344)
  File "zmq/backend/cython/checkrc.pxd", line 25, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/socket.c:10014)
zmq.error.ZMQError: Operation cannot be accomplished in current state
```

Once, I got this error instead:

```
2017-09-20 16:42:57,261 INFO lewis.ChopperEpicsInterface.PropertyExposingDriver: Processed PV updates: State=init
Resource temporarily unavailable (src/signaler.cpp:301)
Aborted (core dumped)
```

If it doesn't crash the first time, try running test.py a couple more times. Careful, on master test.py runs **MUCH** slower! But it doesn't seem to ever crash lewis.

I first saw the error just using `lewis-control` manually, the script is just to make it easier to replicate.

Reply:
I think I've found the problem, there was still a call to `process` in the main thread, I must have overlooked it. I was not able to reproduce the problem after the latest commit - could you please verify?

Also, the script basically finishes instantaneously...it's amazing. I even implemented a request counter on the server side (did not commit it though) because I could not believe it...

Reply:
Ran the script several times, then upped the loop counter to 100,000... still no crashes! Nice catch!